proxy/reverse.go

```diff
@@ -203,3 +212,41 @@ func maybeSetForwardedFor(req *http.Request) {
 	}
 	req.Header.Set("X-Forwarded-For", clientIP)
 }
+
+// recordProxy records proxyID and an endpoint in HTTP headers,
+// in order to detect a proxy loop.
+func recordProxy(req *http.Request, proxyID string, loc url.URL) {
+	pv := proxyID + "=" + loc.String()
+	if prior, ok := req.Header[proxyLoopHeaderKey]; ok {
+		pv = strings.Join(append(prior, pv), ", ")
+	}
+	req.Header.Set(proxyLoopHeaderKey, pv)
+}
+
+// checkProxyLoop detects a proxy loop by passing around proxyID
+// and endpoints. It returns the repeating endpoint string and
+// true if it finds the loop.
+func checkProxyLoop(req *http.Request, proxyID string) (string, bool) {
+	if prior, ok := req.Header[proxyLoopHeaderKey]; ok {
+		for _, v := range strings.Split(prior[0], ", ") {
+			vs := strings.Split(v, "=")
+			if len(vs) == 2 {
+				if proxyID == vs[0] {
+					return vs[1], true
+				}
+			}
+		}
+	}
+	return "", false
+}
+
+// banEndpoint bans the URL from reverseProxy's endpoints.
+func (d *director) banEndpoint(rp string) {
+	d.Lock()
+	d.bannedEndpoints[rp] = struct{}{}
+	d.Unlock()
+
+	// TODO: immediately update endpoints by
+	// calling refresh method.
+	// d.refresh()
```

Comment:
@yichengq I commented this lines out. What do you think?
I agree but if we don't exclude the endpoint here by calling refresh,
we would still have proxy loop until the next refresh call, which is 
by default 30-second.